src/include/mallocMC/reservePoolPolicies/CudaSetLimits_impl.hpp

```diff
@@ -27,26 +27,40 @@
 
 #pragma once
 
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+
 #include "CudaSetLimits.hpp"
 
 #include <cuda_runtime_api.h>
+#include <mutex>
 #include <string>
 
 namespace mallocMC
 {
     namespace ReservePoolPolicies
     {
+        // TODO alpaka
         struct CudaSetLimits
         {
-            static auto setMemPool(size_t memsize) -> void *
+            template<typename AlpakaDev>
+            auto setMemPool(const AlpakaDev & dev, size_t memsize) -> void *
             {
-                cudaDeviceSetLimit(cudaLimitMallocHeapSize, memsize);
+                // FIXME: cannot call cudaDeviceSetLimit() anymore if any kernel has been launched
+                static auto first = cudaDeviceSetLimit(
+                    cudaLimitMallocHeapSize,
+                    memsize); // abusing thread-safe static init for calling
+                              // cudaDeviceSetLimit only once per process
                 return nullptr;
             }
 
             static void resetMemPool(void * p = nullptr)
             {
-                cudaDeviceSetLimit(cudaLimitMallocHeapSize, 8192U);
+                // see:
+                // https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g05956f16eaa47ef3a4efee84563ccb7d
+                // "Setting cudaLimitMallocHeapSize must not be performed after
+                // launching any kernel that uses the malloc() or free() device
+                // system calls"
+                // cudaDeviceSetLimit(cudaLimitMallocHeapSize, 8192U);
```

Comment:
CUDA does not allow us to call this a second time :/ What should we do?

Reply:
Let us add this behavior to the documentation of the policy `CudaSetLimits` and disable the failing test for this policy. If possible write out a message that the test is disabled for policy XY.

Reply:
I reverted to the old behavior of the policy and documented the problem. As for the tests, I could not solve them using a `cudaDeviceReset()` alone. It seems I also need to clear the error of the call to `cudaDeviceSetLimit()` in `resetMemPool()`